census/vanilla-tf/trainer/task.py

```diff
@@ -0,0 +1,338 @@
+  """Sparse feature cross of the feature SparseTensors."""
+  return sparse_feature_cross_op.sparse_feature_cross(
+      feature_tensors,
+      hashed_output=True,
+      num_buckets=num_buckets,
+      hash_key=tf.contrib.layers.SPARSE_FEATURE_CROSS_DEFAULT_HASH_KEY,
+      name=name
+  )
+
+def generate_wide_columns(input_columns):
+  """Generate wide columns by adding feature crosses of SparseTensors."""
+  (age, education_num, capital_gain, capital_loss, hours_per_week,
+   gender, race, education, marital_status, relationship, workclass,
+   occupation, native_country) = input_columns
+
+  # bucketize the age feature
+  # See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/ops/bucketization_op.py
+  age_bucket = bucketization_op.bucketize(age, [18, 25, 30, 35, 40, 45, 50, 55, 60, 65])
+
+  wide_columns = [
+      (sparse_cross([education, occupation], 15 * 16, name='edu_occ'), 15 * 16),
+      (gender, 2),
+      (native_country, 42),
+      (education, 16),
+      (occupation, 15),
+      (workclass, 9),
+      (marital_status, 7),
+      (relationship, 6)
+  ]
+
+  dense_tensors = [sparse_to_dense(col[0], col[1]) for col in wide_columns]
+
+  return tf.concat(dense_tensors, 1)
+
+def sparse_to_dense(sparse_tensor, vocab_size):
+  """Convert the sparse to dense tensor."""
+  dense_tensor = tf.sparse_to_indicator(sparse_tensor, vocab_size)
+  dense_tensor = tf.cast(dense_tensor, tf.int32)
+  return dense_tensor
+
+def read_input_tensor(input_file, skiprows=None):
+  """Concatenate the wide columns to produce a single tensor."""
+  inp, label = read_input_data(input_file, skiprows)
+  in_tensor, label_tensor = generate_input(inp, label)
+  return generate_wide_columns(in_tensor), label_tensor
+
+
+#
+# Function to perform the actual training loop.
+# This function is same for single and distributed.
+#
+def training_steps(session, graph, inputs, labels,
+                   max_steps, train_eval_tensor,
+                   job_name='local', job_id=0):
+  """Run the training steps and calculate accuracy every 10 steps."""
+
+  train_step, eval_step, global_step = graph
+  step = tf.train.global_step(session, global_step)
+
+  (train_inp, train_label), (eval_inp, eval_label) = train_eval_tensor
+
+  while step < max_steps:
+    session.run(
+        train_step,
+        feed_dict={
+            inputs: session.run(train_inp),
+            labels: session.run(train_label)
+            }
+    )
+
+    step = tf.train.global_step(session, global_step)
+
+    if step % 10 == 0:
+      accuracy = evaluate_accuracy(session, eval_step, inputs, labels,
+                                   eval_inp, eval_label)
+      print('[{}/{}]: Step number {} of {} done, Accuracy {:.2f}%'.format(
+          job_name, job_id, step, max_steps, accuracy))
+
+  return train_step
+
+
+#
+# Single and Distributed training functions.
+#
+def training_single(job_dir, max_steps, train_eval_tensor):
```

Comment:
Collapse to single function -- training in single node and distributed node should be more similar and the differences should be keyed off of which task the code is running in -- master or worker.

Workers run the training loop, and log every N steps.
Master is a worker + logs global steps/loss every N steps, produces a summary every N steps, does a checkpoint every M seconds and runs evaluation over that checkpoint, and finally saves a model.